emit_IR.py

```diff
@@ -49,6 +58,8 @@ def visit_SFunction(self, func: sexp.SFunction) -> None:
         )
 
         func.code = emitted_func
+        assert func.name not in self.global_env, (
+            f"Duplicate function name: {func.name}")
```

Comment:
what, you don't like shadowing? :P
(I realize that with our current dynamic binding, shadowing makes the recursion optimization unsound…)

Reply:
Ah, we could avoid this if we made the function's name always available as a local? Then it would be lexically scoped, and also would avoid a `lookup` instruction.

Reply:
Anyway, this is just me thinking, the `assert` is fine and we'll leave it as is.